File: core/providers/json_text.py
"""LLM 文本响应中的 JSON 提取工具（Anthropic / OpenAI 兼容层共用）。"""
from __future__ import annotations


def strip_code_fence(text: str) -> str:
    text = text.strip()
    if text.startswith("```"):
        first_line_break = text.find("\n")
        if first_line_break != -1:  # 掉落 ```json 等语言标记行
            text = text[first_line_break + 1:]
        if text.rstrip().endswith("```"):
            text = text.rstrip()[:-3]
    return text.strip()


def extract_json_object(text: str) -> str:
    """提取第一个完整的最外层 JSON 对象（感知字符串内的花括号与转义）。

    兼容模型在 JSON 前后偶发附加说明文字的情况；无花括号时原样返回。
    """
    start = text.find("{")
    if start == -1:
        return text
    depth, in_string, escape = 0, False, False
    for i in range(start, len(text)):
        ch = text[i]
        if in_string:
            if escape:
                escape = False
            elif ch == "\\":
                escape = True
            elif ch == '"':
                in_string = False
        elif ch == '"':
            in_string = True
        elif ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                return text[start:i + 1]
    return text[start:]

File: src/providers/jsonText.ts
/** LLM 文本响应中的 JSON 提取工具。 */

export function stripCodeFence(text: string): string {
  let t = text.trim();
  if (t.startsWith("```")) {
    const firstBreak = t.indexOf("\n");
    if (firstBreak !== -1) t = t.slice(firstBreak + 1); // 掉落 ```json 等语言标记行
    if (t.trimEnd().endsWith("```")) t = t.trimEnd().slice(0, -3);
  }
  return t.trim();
}

/** 提取第一个完整的最外层 JSON 对象（感知字符串内的花括号与转义）。 */
export function extractJsonObject(text: string): string {
  const start = text.indexOf("{");
  if (start === -1) return text;
  let depth = 0;
  let inString = false;
  let escape = false;
  for (let i = start; i < text.length; i++) {
    const ch = text[i];
    if (inString) {
      if (escape) escape = false;
      else if (ch === "\\") escape = true;
      else if (ch === '"') inString = false;
    } else if (ch === '"') inString = true;
    else if (ch === "{") depth += 1;
    else if (ch === "}") {
      depth -= 1;
      if (depth === 0) return text.slice(start, i + 1);
    }
  }
  return text.slice(start);
}
